### ISSUE TITLE
Feature Request - Method to Purge a document

### ISSUE THREAD

**Issue by jpgreen** [pouchdb/pouchdb#802]
Feature Request - Method to Purge a document

Note: Purge is different from remove (or delete). It is as though the database had never heard of document at all. 

Info: [http://wiki.apache.org/couchdb/Purge_Documents](http://wiki.apache.org/couchdb/Purge_Documents)

**Comment by doublerebel** [comment]
@ddouglascarr your purge addition looks good, is there anything missing?  I'm looking for this feature also and am willing to help.

**Comment by ddouglascarr** [comment]
Hi @doublerebel. Sorry, I missed you mention in my events feed. 

There's quite a bit to do, and any help would be fantastic.
* I've just committed the websql adapter stuff, but that leaves the indexDB and levelDB adapters, which I have no experience with.
* purge_seq is not currently getting recorded. (This will probably involve modifying PouchDB.prototype.info()). This is my next mission.
* The view and find plugins will need updating. (I'm going to make a comment about this shortly)
* The above may mean this needs to be turned into a plugin. (Again, see next post).

**Comment by ddouglascarr** [comment]
Implementing this feature will involve breaking backwards compatibility (and bumping the major version). Because of this, I think that it should perhaps be a plugin instead.

The problem is that any indexing plugins (pouchdb-view, pouchdb-find, etc) will need to check the databases purge_seq, react, and re-index from scratch if the purge sequence has advanced more than one since their last indexing event.

This is detailed in the couchdb documents here: http://docs.couchdb.org/en/latest/api/database/misc.html#updating-indexes

I think what I've done so far could be made into a plugin pretty easily. I don't know about the purge sequence though. I'm planning on looking into that next. Having it be a plugin would have the advantage of hiding the functionality from new users, who tend to struggle to differentiate between purge and remove (with potentially disastrous consequences). 

What are people's thoughts here?

**Comment by stenit** [comment]
I'd like to have the plugin approach due to the "urgency" of this feature. Later on, this could be part of a new major version. To bridge this time, for me a plugin sounds like a good solution.

**Comment by nolanlawson** [comment]
I propose we resolve this problem by adding a new event to the `changes()` API: `'purge'`. Then the mapreduce plugin can simply listen for purge events (as it's already listening to changes) and delete as necessary. It'll just be understood that http adapters don't emit this event. @daleharvey  @calvinmetcalf thoughts?

**Comment by nolanlawson** [comment]
Actually I take it back; that's a silly idea. What if the listener isn't listening for that event? It'll never know the document was purged.

I have a better idea: this should wait until we've implemented native secondary indexes, at which point it should be easier to implement, without the inelegant hacks needed for `dependentDbs` etc.

**Comment by ddouglascarr** [comment]
@nolanlawson The way I was intending on doing it is the way couchdb does it. There's a purge_seq on the couchdb database and it caches the last purge response. The view server checks if the purge has incremented. If it has, it reads the cache and adjust. If it's incremented more than 1 it re-indexes from scratch. 

Not at all ideal, but quite simple to implement. The user just has to query all affected views after every purge. I really like what you're suggesting with the native secondary indexes, and it'd be great to do without. 

Could you have a look at what I've done with purge so far (links above)? Is that in the right direction?

**Comment by nolanlawson** [comment]
@ddouglascarr Unfortunately that's not quite right... modifying `dist/pouchdb.js` won't do anything because it's a generated file. Also this should probably be fixed in the individual adapters `idb.js`, `leveldb.js`, etc., since otherwise there's no way to truly "delete" the data from the underlying datastore.

I'm committed to getting this feature implemented, but right now I'm swamped with other things, and I only work on PouchDB on the weekends. In general, I still think this feature should probably wait for native secondary indexes, since otherwise it's going to be a pain to remove data that's been created by map/reduce (as pointed out by others).

**Comment by ddouglascarr** [comment]
@nolanlawson, thanks for having a look. Sorry I wasn't clear, there's a bunch of commits on my branch, not just the last one. The changes in dist/pouchdb.js are just me forgetting to commit those from before, the changes have been made to the source files.

I haven't done the idb, leveldb stuff yet, just websql (which is what I'm most familiar with). I'm pretty sure it works, and i've included a purge_seq as well.

**Comment by ddouglascarr** [comment]
@nolanlawson. Also, the revision tree would be the same in the system you're proposing right? Most of what I've done is just a standalone method to alter the revision tree.

**Comment by EduceHealth** [comment]
@ddouglascarr ,@nolanlawson,  Any progress on this front? I'm eagerly waiting to implement some kind of purge before I cripple people's phones entirely!@ddouglascarr ,@nolanlawson,  Any progress on this front? I'm eagerly waiting to implement some kind of purge before I cripple people's phones entirely!

**Comment by Grokling** [comment]
@ddouglascarr @nolanlawson , Any progress on this front? I'm eager to implement some kind of purge before I cripple people's phones entirely.

**Comment by ddouglascarr** [comment]
@Grokling, hi. I actually have something I'm using. It only works on websql though, and you have to query all your views every time you use purge. I didn't want to go to the hassle of doing the other databases as a better solution is in the pipeline. You can see it at my github. pouchdb#802 and mapreduce#pouchdb802. I'm going to be doing some more testing in the next couple of days, so keep an eye on it.

If you're using idb, it wouldn't be that hard to implement there. Most of the code is in core, and the tests are written. Just need to write the adapter (for pouchdb and pouchdb-mapreduce ).

**Comment by ssured** [comment]
What is the status of this? I'd love to help, but do not know where to start. Right now my users destroy the DB and the sync again to get rid of old stuff. Purging would be a much better solution for me.

**Comment by nolanlawson** [comment]
This requires some subtle fixes to map/reduce and to the adapters, and is currently on hold.

**Comment by chrismbeckett** [comment]
Wow, after months of working to get Couchbase Sync Gateway working with PouchDB this issue just trashed me. So it starts with the inability of Couchbase Sync to inform the client of documents removed (not changed or deleted, just removed due to security) from a sync profile.(https://github.com/couchbase/sync_gateway/issues/264). But... even if they could, I would not be able to actually purge the documents the local PouchDB.

As suggested by @ssured, this seems to suggest that if you have any sort of multi-user security model for your application, anytime anything in a security model changes, all client databases have to be deleted and completely re-synched?

Please... please, make this a high priority item to fix in the next release! This is a brutal limitation.

**Comment by nolanlawson** [comment]
@chrismbeckett I understand your frustration. A couple quick points:

1. Couchbase Sync Gateway support has never been perfect. We don't even test it in integration tests. ([We've always been honest about this.](http://pouchdb.com/adapters.html#pouchdb_over_http)) Basically CSG only implements a subset of the CouchDB API, so we struggle to support it 100%. PouchDB Server, CouchDB 1.x, CouchDB 2.x, and Cloudant are fully supported. 
2. I realize it's an important feature, but this is an unfunded open-source project. Speaking for myself (dunno about the other contributors), this is a project that I work on during the weekends and evenings, instead of spending time with my family. We'll get to it when we get to it. Thanks for understanding. :)

**Comment by froatsnook** [comment]
I'm a brand new pouchdb user and I am also worried about infinite database growth.  Instead of purge, would it be possible to get around this issue by re-using IDs of deleted items, and then compacting?

**Comment by Grokling** [comment]
@froatsnook, if you're deleting (which is an entirely different animal to purging) this is possibly a non-issue for you.

Purge is primarily aimed at the scenario where documents are replicated from somewhere else to a client, but are then no longer needed on that client - but at the same time should remain valid and present on the 'somewhere else'. 

Think of it as "forget on this client", which is quite distinct from "delete from the distributed database".

**Comment by froatsnook** [comment]
@Grokling You're right -- I am in the wrong thread!

**Comment by nkoder** [comment]
For me lack of `purge` is kind of a blocker: I cannot ensure that database usage in a browser on mobile device will not grow up to its limit in a longer time. Moreover I have no other replicas to synchronize with so replication issues are not issues in my case.

**Comment by nickcolley** [comment]
Hey @nkoder,

We're aware that purge is a super important feature for the same reasons you've raised, however this is a complex feature to ship with the current architecture in the adapters currently.

There's ongoing work in https://github.com/pouchdb/pouchdb/pull/4984 to explore a simpler underlying data structure that could make implementing purge much easier.

One alternative at the moment is to create a new database for the data you want to keep and delete the previous, clearly not ideal.

Lastly, you've mentioned you're not using PouchDB to replicate at all, if syncing is not a requirement of your app you could consider something like [localForage](https://github.com/mozilla/localForage) in the meantime.

**Comment by chrismbeckett** [comment]
@nkoder - take a look at ForerunnerDB.

**Comment by nkoder** [comment]
@nickcolley Thanks for localForage recommendation. It looks like the minimal tool needed to cover all out requirements (eg. being a facade to IndexedDB, WebSQL and LocalStorage)

**Comment by mjosephd** [comment]
@chrismbeckett We also are using couchbase sync gateway and ran in to the same problem, did you guys find any workaround for this or rather how did you accomplish this?

**Comment by mjosephd** [comment]
@chrismbeckett We also are using couchbase sync gateway and ran in to the same problem, did you guys find any workaround for this or rather how did you accomplish this?

**Comment by chrismbeckett** [comment]
@mjosephd I ended up after many twists and turns switching to Parse on the server (after they added LiveQuery) with LocalForage and TaffyDB on the client. I wrote my own logic to sync.

**Comment by ssured** [comment]
It depends on your app, but for my usage it's fine to sometimes purge a whole DB and just sync again. I did have worries about purging, but in the end for my use case it was not needed, DB's did not grow as fast as I expected. My way of syncing is using https://github.com/conor-mac-aoidh/express-pouchdb-replication-stream which converts a whole DB to a single downloadable file. Works great.

**Comment by kripper** [comment]
PouchDB without purge, is like a car without wheels.
Yes, it syncs, but just until it crashes.
Deleting databases and resyncing is not an option.
Any chance to have sygi's patches merged?

**Comment by episage** [comment]
+1

**Comment by nolanlawson** [comment]
Folks, I understand you want this feature, but it's currently blocked on `idb-next` work which is still underway (https://github.com/pouchdb/pouchdb/issues/5207). It's a very complicated feature to implement, and we want to make sure we do it right.

In the meantime, the workaround is to create a second local database, do a filtered replication to it, and then delete the original database. Thanks for understanding!

**Comment by kripper** [comment]
Seems like a major redesign is going on and there are concerns to not break backwards compatibility which will probably slow down development of this critical features.

In case you decide to create a branch (breaking backwards compatiblity to ease development), we would be very happy to do exhaustive testing of pouch's core features (read, delete, push, pull, sync and *purge*).

**Comment by kripper** [comment]
Regarding the workaround of creating a second database, it's not possible if the mobile device is running out of space (that's why we need to purge).

And recreating databases + resync'ing is also problematic because running out of space also means that we have a big number of documents and resync'ing will take too long.

**Comment by akatov** [comment]
@kripper - maybe you can create the temporary database using the in-memory adapter
https://www.npmjs.com/package/pouchdb-adapter-memory

**Comment by kripper** [comment]
@akatov I don't think so, because we develop for smartphones and the database size exceeds RAM. Thanks anyway.

**Comment by natew** [comment]
Just wanted to chime in and say I'd be pro a private and non-guaranteed build that was meant for people doing greenfield projects, if that would help get it out quicker and allow testing. If anything it would give us a better place to track the status of this.

**Comment by federicobouzas** [comment]
I need something like this too!

**Comment by federicobouzas** [comment]
I would love something like this too! Thanks!!

**Comment by nikhilagw** [comment]
+1

**Comment by 4F2E4A2E** [comment]
+1

**Comment by kripper** [comment]
Please fix. This feature is critical.

**Comment by episage** [comment]
+1

**Comment by hermitdemschoenenleben** [comment]
+1

**Comment by natew** [comment]
Figured some part of Copenhagen ethics tips me to reply here. I feel
pouch/couch are about a step away from being a perfect solution to their
problems each. On the pouch side this is the big one, the other recent
great pieces that have fallen in place. On Couch, from the looks of the PRs
it’s on the right track. Would love to see this land this year.

**Comment by kripper** [comment]
Today another customer (a big hospital in Chile) started to complain about this issue.
Deleting the data from the mobile phone and resync'ing is not a solution, because it's a big database and the downtime is unacceptable.

I agree with @cepm-nate that other developers should be warned of this important limitation before starting to use PouchDB, because it is a timebomb. Please link this issue in the docs.

What should we do know? Rewrite the app to use another framework? Which one? Does anybody know of an easy replacement?

The biggest problem I see is that after some years of the development of the app, there is no budget to rewrite it.

**Comment by cepm-nate** [comment]
@kripper, our system has a few layers, like this: 
MSSQL <-> custom live sync javascript <-> CouchDB <-> PouchDB <-> LokiJS <-> App GUI.
 - Actually, the App GUI saves changes directly to PouchDB, which then update's LokiJS so that queries are lightning fast. (Queries against PoucDB are slow, so we mirror the data into LokiJS for speed).

Yes ,there are drawbacks! The main one is keeping the PouchDB and LokiJS in sync, but that is not related to the topic at hand.

My next approach will be like this:
MSSQL <-> Custom Node API <-> LokiJS.

The custom node API can reuse much of the saving/getting code from the LIVE SYNC JAVASCRIPT we wrote initially, as it is doing the converting to and from our master MSSQL DB.

LokiJS has some adapters for persisting the data onto the disk. It even has event handlers to fire when data is changed. You can also run LokiJS on the backend server, but you would still need to write all the syncing middleware including the API on the server.

FireBase and Realm are a couple Document DB's that I ran into that support offline-first functionality. Perhaps you could start with them for a more integrated solution?

**Comment by daleharvey** [comment]
The workaround doesn't require resyncing from the remote, you can do a local sync. And prodding OSS Devs to implement features because your paying customers need them isn't going to gain much sympathy. I plan on implementing this as part of idb-next, if that isn't soon enough for someone a financial interest they can always contribute it themselves

**Comment by cepm-nate** [comment]
So on the local sync (phone) side, are you suggesting something like this?
 - Halt replication between old DB and server
 - Create a new replicated DB from the old DB (filter out deleted, latest rev only)
 - Delete old DB
 - Rename new DB to old DB name
 - ReInit replication between new DB and same old server?

Wouldn't all the old revision history from the server still get sent to this new DB? What if the local device is already very low on space?

**Comment by marten-de-vries** [comment]
@cepm-nate Purging would have the same problem. From [the CouchDB docs](http://docs.couchdb.org/en/2.0.0/api/database/misc.html):

> If you are replicating between databases and have deleted a large number of documents you should run purge on each database.

So, the answer is to do the same (or use ``_purge``, as it actually exists there) server-side. As for a device being very low on space, I guess your only alternative currently is to throw away data and redownload it. Keep in mind that recovering from such a situation is a useful feature anyway, as browsers may clear IndexedDB for whatever reason they like, whenever they like.

**Comment by daleharvey** [comment]
> So on the local sync (phone) side, are you suggesting something like this?

I would keep a rolling database, switching once per month / week / day depending on your usage needs, replicating does not transfer the entire revision history (which either way, can be limited using revs_limit) and can filter out deleted documents. If the space is too limited for that, then I would implement purge

I cant think of a particularly sensible place to put a notification about the things we dont implement (of which there are lots), purge may be worth a mention in the faq, but as will be a common theme here, if you want something done, its an OSS project, make a PR

**Comment by kripper** [comment]
Can someone please comment on this commit implementing purge:
https://github.com/sygi/pouchdb/commit/116594f547957322126b8375b897566bed6d78ce

**Comment by marten-de-vries** [comment]
@kripper The last said on it is https://github.com/pouchdb/pouchdb/issues/802#issuecomment-57162970 and the further reactions. In short, it seems it never supported indexes and would not merge anyway since it's a couple of years old. Also, the ``idb-next`` push will render at least large parts of its code obsolete as the underlying data store schema changes.

**Comment by alxndrsn** [comment]
@daleharvey would it be practical and useful to implement the rolling databases as a pouchdb plugin?  With some kind of interface like:
```
db.rollover()
  .then(info => console.log('Rollover completed.', info))
  .catch(e => console.log('Something went wrong.', e);
```

**Comment by alxndrsn** [comment]
> And prodding OSS Devs to implement features because your paying customers need them isn't going to gain much sympathy.

Are there any individuals or orgs that do paid consulting/development on PouchDB?  Might be useful to know for organisations with financial pressure for new features or bugfixes.

**Comment by daleharvey** [comment]
> Can someone please comment on this commit implementing purge: sygi/pouchdb@116594f

Its a good basis if someone wanted to pick it up and get it mergeable (particularly the tests), as said it will have bitrotted a lot by now and didnt implement purge_seq at the time

> @daleharvey would it be practical and useful to implement the rolling 
> databases as a pouchdb plugin? With some kind of interface like:

Not sure that a plugin is needed, a function like https://pouchdb.com/2018/01/23/pouchdb-6.4.2.html that did some trickery with the database names seems fairly simple, be happy to add it to the docs or have a blog post to point to about it

> Are there any individuals or orgs that do paid consulting/development on 
> PouchDB? Might be useful to know for organisations with financial pressure 
> for new features or bugfixes.

The folks at https://neighbourhood.ie/ are usually the first place I point people, there are quite a few freelancers etc who are familiar with PouchDB that have done consulting in the past, the mailing list is a reasonable place to look or happy to retweet people looking for that type of work done

**Comment by kripper** [comment]
This bug is critical.
Please fix or place a bounty:
https://www.bountysource.com/issues/796221-feature-request-method-to-purge-a-document

**Comment by kacerik** [comment]
hey, anyone willing to fix this for 250usd bounty?

**Comment by kripper** [comment]
Maybe you should place the bounty. As far as I know, you can request the money back if no one applies.

**Comment by kripper** [comment]
Please fix / add more bounties.
This feature is mandatory.

**Comment by dave-irvine** [comment]
I bountied this for $130, looks like its tough to fix but having the issue since 2013 is not great :/

**Comment by mz2** [comment]
Happy to chip in a $150 also... (just not on bountysource which seems to grab funds upfront and then grab 10% if you want a refund!)

**Comment by kripper** [comment]
Great!
Are you sure about the 10%?
I didn't have problems refunding last time I tried (via E-mail).

**Comment by mz2** [comment]
That's what the FAQ states. Anyhow, to stay focused on the task, I'd also be happy to try solve this with @will118 as we really need this feature. 

@daleharvey, could you help a prospective entry level contributor with some more detail on your comments above:

- Taking https://github.com/sygi/pouchdb/commit/116594f by @sygi as a basis, could you point at the kind of bitrot that we'd expect to be dealing with? Any dummy level pointers of changed APIs or relevant changed implementation details are helpful.
- By purge_seq do you mean that the sequence number should be incremented by purge calls? What would need to be stored against this sequence number?
- Regarding the test coverage, are the tests missing scenarios that should be tested?

**Comment by daleharvey** [comment]
> Any dummy level pointers of changed APIs or relevant changed implementation details are helpful.

There are a few minor things it looks like, the http api is changed but there are plenty of examples in https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-adapter-http/src/index.js#L705 to work from, the bulk of the work is on the rev_tree which for the standard adapter is unchanged.

> By purge_seq do you mean that the sequence number should be incremented by purge calls? 
> What would need to be stored against this sequence number?

So the purge_seq is used to notify `.query` calls that a purge has happened so it can be sure to not include purged results, thats about as much as I know I havent looked much more into it

> Regarding the test coverage, are the tests missing scenarios that should be tested?

The above scenario is the main one missing I think

**Comment by daleharvey** [comment]
And happy to support the work either way, however the main reason this has been left open for so long is the hassle of the interaction between `purge` and `query`, we cant release a `purge` in the current adapter that breaks with map reduce, but a purge implemented in `idb-next` (https://github.com/pouchdb/pouchdb/issues/5207) does not need to worry about map reduce and would be far easier and faster to implement

We only have one blocker for releasing `idb-next` (https://github.com/pouchdb/pouchdb/issues/6543) and there is good momentum and enthusiasm for doing so, I would seriously consider putting your efforts towards idb-next

**Comment by snowyu** [comment]
Why close it ? This is an important issue.

**Comment by alxndrsn** [comment]
@snowyu it's a bot.  But your reply re-opened the issue :slightly_smiling_face:

**Comment by plinss** [comment]
I presume you're aware, but just in case, I hope the PouchDB solution here works with CouchDB's new clustered purge feature: http://docs.couchdb.org/en/stable/cluster/purging.html#cluster-purging

**Comment by jamesmfriedman** [comment]
Definitely in need of this feature as well... A few things

I was able to hack in "live filtering" by tearing down the old sync and setting up a new one whenever the filter changes. This is less than ideal, but it works.

I am now working on purging local data! My app deals with lots of attachments, so it is critical that I be able to clean them up when they're not being used.

The other option is to rebuild this whole thing without replication, but I would be adding a ton of complexity for something that is 99% working, just missing the Purge feature.

My current plan is to destroy the local database to clear everything from memory :/. This sucks.
This issue has been open for 5 years now, any chance it's close to done?

**Comment by karlwestin** [comment]
Yes the PouchDB contributors are quite aware of the work that's been done to implement `purge` across different nodes in couchdb ;)

Here's some good info on the complexity of `purge` and why it took such an effort to bring the Clustered Purge into CouchDB https://twitter.com/janl/status/1072442448893358081

PouchDB is an open source effort, feel free to hit us up in the chat or here if you'd like to start implementing this.

**Comment by jamesmfriedman** [comment]
I'm no stranger to OSS, have a few of my own ;). Sorry, didn't mean to come across as critical. Better translation...
- I have something I need to ship that needs this. Is it a soon or someday thing?
- If it's someday (which it sounds like it) I'll go with the workaround.

PouchDB has been a joy to work with, thanks for the hard work.

**Comment by karlwestin** [comment]
cool :)
unfortunately there's noone working on this at the moment, so probably you're better off going for the work around

Implementing Purge in PouchDB is something of a massive effort, and anyone who can contribute anything toward that goal will be a real community hero!

**Comment by jamesmfriedman** [comment]
More of a curiosity then, does PouchDB know if I manually delete an underlying record in the local DB? This seems like it would fulfill the local client issues and provide a more direct workaround.

**Comment by dave-irvine** [comment]
Yeah I also needed this in order to ship something, posted a bounty for it but still nothing unfortunately so had to go with the workaround of deleting the entire db and recreating.

**Comment by daleharvey** [comment]
So once we land https://github.com/pouchdb/pouchdb/pull/7547 then in in `pouchdb.next` this is going to be a reasonably small task for that adapter

The complexity was mostly in implementing it for map reduce, I am happy to not support purge + map reduce but if thats a thing someone wanted then yeh would probably be a bit away

**Comment by daleharvey** [comment]
> More of a curiosity then, does PouchDB know if I manually delete an underlying record in the local DB? This seems like it would fulfill the local client issues and provide a more direct workaround.

PouchDB doesnt know, but if what it expects to read no longer exists then it can fail in undetermined ways, thats basically corrupt data at that point

**Comment by jamesmfriedman** [comment]
Cool, that’s very helpful! Gives me a place to start exploring.

**Comment by SCdF** [comment]
FWIW our temporary hack for purging is just replacing a document we want to purge with:

```json
{
  "_id": "foo",
  "_rev": "1-bar",
  "_deleted": true,
  "purged": true
}
```

And then:
 - Compacting after running purges (we purge based on rules that run every so often and may "purge" many docs in one go)
 - Changing our upwards replication from PouchDB to CouchDB is filtered, and doesn't replicated up documents that fit this pattern
 - For extra safety, changing CouchDB's `validate_doc_update` function to reject documents with the `purged` flag

It's not purging, but logic-wise it presents as we'd want purging to (data still exists on server, but is gone from phones), and it does improve performance somewhat.

**Comment by SCdF** [comment]
We'll get there, stale[bot], we'll get there.

**Comment by jbotte** [comment]
My unique scenario is I have anonymous users filling out surveys at events. Once the data is replicated to online couches I no longer need the data on the device. If I one-way sync can I somehow mark any items that have successfully replicated for deletion and if the entire local dB is marked as synced I just delete and recreate the pouch dB?

**Comment by SCdF** [comment]
@jbotte it's hard to tell from what you're written there, but it sounds like you're just using PouchDB as an offline-safe queue to push data to a server?

I'm not sure if PouchDB is the right tool for that honestly, though I'm not sure what an alternative would be (I am not up to date with the space though). 

Regardless, you can achieve something like that with PouchDB yeah, no purge required. Replication is an event emitter, either continuous or one-off, and so you can use those events to work out the replication state and do things accordingly. If you want to delete as you go you can filter upward replication so that it doesn't replicate deletes upward

**Comment by jbotte** [comment]
Yes that is an accurate description “offline save queue”

I agree pouch may not be the best tool for this, I’m researching other alternatives but like how easy pouch has been, plus I may have a second content database that is two way synced and liked the idea of not adding more frameworks than necessary if I could accomplish both with pouch. 

Thanks for the advice.

**Comment by garrensmith** [comment]
I wouldn't use replication and purge for this. Rather just store the data locally in a db. Once you are back online then fetch the local document from pouchdb and save it to the remote db. Then delete the document locally. If you local db gets a bit big and you have no unsaved local docs, delete it and create a new one

**Comment by jbotte** [comment]
thinking about a potential solution to purge...

what about something like this firing each app launch?

`var activeDB = new PouchDB('activeDB');
var cleanedDB = new PouchDB('cleanedDB');
var originalTableSize; 
var cleanedTableSize;

activeDB.info().then((result)=>{
    originalTableSize = result.doc_count;
    activeDB.replicate.to(cleanedDB, {filter : function(doc) { if( doc._deleted ) { return false; } else { return doc; } }}).on('complete', function (result) {
        cleanedDB.info().then((cleanedResult)=>{
            cleanedTableSize = cleanedResult.doc_count;
            if(cleanedTableSize == originalTableSize) {
                console.log('well sinice things match destory old');
                activeDB.destroy().then(()=>{
                    console.log('now that old is destroyed copy it back to active');
                    activeDB = new PouchDB('activeDB');
                    cleanedDB.replicate.to(activeDB, {filter : function(doc) { if( doc._deleted ) { return false; } else { return doc; } }}).on('complete', function (result) {
                        console.log('we copied clean back to active');
                        console.log('destroy clean');
                        cleanedDB.destroy().then(()=>{
                            console.log('we deleted the clean');
                            console.log('now we will kick off the real sync between pouch and couch');
                        });
                    });
                })
            }
        });
    });
});


`

**Comment by SCdF** [comment]
If that works for your application / use case you can solve it manually that way. However, it has many downsides you should be aware of:

 - you lose any _local docs, which among other things may affect replication (eg to a CouchDB on a server) and cause it to start again, depending on configuration.
 - it’s not going to be fast to transfer the data there and back, depending on data sizes
 - all views will have to re-index, which is really slow, again depending on data sizes

**Comment by jbotte** [comment]
SCdF, good points I hadn't tested speed or worked with _local docs yet. Thanks for bringing those caveats to my attention. I may be able to remove that second replicate back by just making cleaned db the active one and keeping the name of active db in local storage as a variable that increments every time the db is cleaned.

**Comment by BananaAcid** [comment]
So what is the current state? Is that possible by now? (just purging like pouchdb never heared of the doc)
It seems like, there are solutions to [purge-on-delete](https://www.npmjs.com/package/pouchdb-purge-on-delete) but not `purge-to-forget`, like this topic was about

**Comment by kripper** [comment]
@SCdF, are your still using a hack?

@BananaAcid, check:
https://github.com/medic/cht-core/issues/5048

**Comment by SCdF** [comment]
@kripper / @BananaAcid effectively yes. We are now doing something much more complicated now, which is intertwined with our custom internal solution to make filtered replication faster.

The core concept though, in relation to purging, is very easy to replicate. There isn't really "code" for this, it's just the convention of using a flag to imply it's been purged locally, and not allowing those deletes to be replicated upwards.

**Comment by BananaAcid** [comment]
Some update from the official side on the topic?

**Comment by ChrisCross3** [comment]
> @kripper / @BananaAcid effectively yes. We are now doing something much more complicated now, which is intertwined with our custom internal solution to make filtered replication faster.
> 
> The core concept though, in relation to purging, is very easy to replicate. There isn't really "code" for this, it's just the convention of using a flag to imply it's been purged locally, and not allowing those deletes to be replicated upwards.

Is there an update on this?
Like the previous speakers, I'm also in need for a purge-to-forget on client solution.

**Comment by Kipsanga-Gregory** [comment]
2022?

**Comment by bean5** [comment]
Please reopen.

**Comment by bean5** [comment]
@sygi Can you share the code you have for this?

**Comment by sygi** [comment]
bean5, I last worked on this issue more than 8 years ago. As far as I can see, the code I wrote can be referenced here: https://github.com/sygi/pouchdb/commit/116594f547957322126b8375b897566bed6d78ce

but I very much doubt you can recover much from it.
